internal/config/config.go

```diff
@@ -0,0 +1,145 @@
+	LDAPEnabled    bool              `toml:"ldapEnabled" validate:"omitempty"`
+	Default        authDefaultConfig `toml:"default" validate:"omitempty"`
+	LDAP           *authLDAPConfig   `toml:"ldap" validate:"omitempty"`
+}
+
+type authDefaultConfig struct {
+	User     string `toml:"user" validate:"required"`
+	Password string `toml:"password" validate:"required"`
+	Visitor  bool   `toml:"visitor" validate:"omitempty"`
+}
+
+type authLDAPConfig struct {
+	Host   string               `toml:"host" validate:"ip|hostname|fqdn"`
+	Port   int                  `toml:"port" validate:"gte=1,lte=65535"`
+	TLS    authLDAPTLSConfig    `toml:"tls" validate:"required"`
+	Bind   authLDAPBindConfig   `toml:"bind" validate:"required"`
+	Search authLDAPSearchConfig `toml:"search" validate:"required"`
+}
+
+type authLDAPTLSConfig struct {
+	Enabled              bool     `toml:"enabled" validate:"omitempty"`
+	SkipCertificateVerif bool     `toml:"skipCertificateVerif" validate:"omitempty"`
+	ThrustedCertificates []string `toml:"thrustedCertificates" validate:"required_without=SkipCertificateVerif"`
+}
+
+type authLDAPBindConfig struct {
+	UserDN   string `toml:"userDN" validate:"required"`
+	Password string `toml:"password" validate:"required"`
+}
+
+type authLDAPSearchConfig struct {
+	Base           string `toml:"base" validate:"required"`
+	Filter         string `toml:"filter" validate:"required"`
+	OwnerGroupDN   string `toml:"ownerGroupDN" validate:"omitempty"`
+	VisitorGroupDN string `toml:"visitorGroupDN" validate:"omitempty"`
+	LoginField     string `toml:"loginField" validate:"required"`
+}
+
+var config tomlConfig
+
+func newTomlConfig() tomlConfig {
+	return tomlConfig{
+		Version: 1,
+		Serve: serveConfig{
+			Port:    8080,
+			Address: "0.0.0.0",
+		},
+		SQL: sqlConfig{
+			Driver:           "sqlite3",
+			ConnectionString: "file:/srv/shiori/shiori.db",
+		},
+		Auth: authConfig{
+			DefaultEnabled: true,
+			SQLEnabled:     true,
+			LDAPEnabled:    false,
+			Default: authDefaultConfig{
+				User:     "shiori",
+				Password: "gopher",
+				Visitor:  false,
+			},
+		},
+	}
+}
+
+func parse(r io.Reader, cfg *tomlConfig) (*tomlConfig, error) {
+
+	if cfg == nil {
+		cfg = &config
+	}
+
+	if _, err := toml.DecodeReader(r, cfg); err != nil {
+		return nil, err
+	}
+
+	validate := validator.New()
+	validate.RegisterStructValidation(authConfigStructLevelValidation, authConfig{})
+
+	if err := validate.Struct(cfg); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func authConfigStructLevelValidation(sl validator.StructLevel) {
+
+	ac := sl.Current().Interface().(authConfig)
+
+	if ac.LDAPEnabled && ac.LDAP == nil {
+		sl.ReportError(ac.LDAP, "ldap", "ldap", "auth.ldap conf missing", "")
+	}
+}
+
+// Config return a pointer to the parsed config
+func Config() *tomlConfig {
+	return &config
+}
+
+func init() {
+
+	cWarning := color.New(color.FgHiYellow)
+
+	config = newTomlConfig()
+
+	file, err := os.Open("shiori.toml")
```

Comment:
I don't think it's good to parse the config file in `init`. First, IMHO error prone method like `os.Open` shouldn't be run on init. Second, since it in init, we can't give the user freedom to specify custom name or path for the config file.